### DISCUSSION TITLE
Include URL percent-encoding?

### DISCUSSION THREAD

**Discussion by KorbisInc** [curl/trurl#219] [General]
## For anyone finding needing this

Read https://github.com/curl/trurl/discussions/219#discussioncomment-6073943 :slightly_smiling_face: 

Original text:

----

Is it too out-of-scope to wish for encoding and decoding URL percent-encoding? A quick search shows there are dozens of code-golfed one liners in the wild, but nothing as simple as `echo $URL | trurl --%-decode`... (It's nice that % isn't a special character in bash...)

A problematic example is trying to use curl on something like `curl https%3A%2F%2Fsecure.gravatar.com%2F` - curl will report a bad hostname. It'd be great if anyone could do `trurl --%-decode https%3A%2F%2Fsecure.gravatar.com%2F | curl`

**Comment by KorbisInc** [comment]
Seeing: https://github.com/curl/trurl/issues/200#issuecomment-1555928544 "**trurl only outputs valid URLs**" - I believe then only `--%-decode` should be supported.

Edit: nevermind, percent-encoded URLs are valid! https://github.com/curl/curl/issues/7830

**Reply by emanuele6** [reply]
I don't see how that is relevant

**Reply by KorbisInc** [reply]
Well if one of turl's design philosophies is "only output valid URLs", then percent-encoded URLs being output would be invalid - that is, until I've just learned they are technically valid too! So nevermind what I said :)

**Reply by emanuele6** [reply]
> Well if one of turl's design philosophies is "only output valid URLs"

It is not really a design philosiphy... trurl by default normalises URLs, you can modify change the style of the URLs a bit with options like `--default-port`, `--keep-port`, `--punycode`, but it only outputs valid URLs in this mode, because that is what it does.
You can print any arbitrary line composed of subparts of the URL with `-g`/`--get`, and you can make it output the parsed URLs as JSON with `--json`.

But what does it even mean to output a "percent decoded URL"? I don't understand what you want here.

Just treat the whole URL as a raw string instead of a URL, and ouptut it as a raw string with % codes decoded?

I don't understand what is the point of using a URL parser like trurl if you don't care about parsing, validating, normalising URLs, or extracting a specific part from a URL, and you just want to percent decode an arbitrary string.

**Reply by KorbisInc** [reply]
You're right, I was focused on the fact I'm dealing with an arbitrary string that just happens to look like a URL.

I suppose then I'll just write a tiny program for this decoding. It would be nice though if there were a de-facto standard tiny tool for it somewhere :smiling_face_with_tear: 

Thank you for taking the time to explain the boundaries of trurl to me. :slightly_smiling_face: Have a great Friday!

**Reply by emanuele6** [reply]
@LeeKorbis

> A problematic example is trying to use curl on something like `curl https%3A%2F%2Fsecure.gravatar.com%2F` - curl will report a bad hostname. It'd be great if anyone could do `trurl --%-decode https%3A%2F%2Fsecure.gravatar.com%2F | curl`

Ok, that example helps! Now I understand more what you want. Your `$URL` is not a URL itself, but it is a string that when percent decoded is a URL. 

If that URL comes from a query in the URL, you can just use something like, that is fine:
```console
$ trurl -g '{query:redir}' 'http://example.org/?redir=https%3A%2F%2Fsecure.gravatar.com%2F'
https://secure.gravatar.com/
```

If it is just a string, and not a part of a URL, and you must use trurl because you cannot use any other tool to urldecode your string so you have no other option, you could use an hack like:
```console
$ trurl -s 'fragment:=https%3A%2F%2Fsecure.gravatar.com%2F' -g '{fragment}' localhost
https://secure.gravatar.com/
```
(note `{:fragment}` and `{fragment}` have accidentally been swapped in `trurl 0.7`, so for trurl 0.7 specifically that should be `{:fragment}`, but that is a bug)

**Reply by KorbisInc** [reply]
Haha, your first example is the exact situation I'm in. To be even more specific, I was manually collecting image examples off of Yandex and put the URLs into a text file, and found that `img_url` held the original URL to the images. A couple of editor kung-fu moves later I had just the percent-encoded URLs but saw that Chromium, then wget, then finally curl couldn't download from them.

`-g '{query:img_url}'` will resolve this problem quite easily :sunglasses: 

Thank you again and stay awesome

**Reply by emanuele6** [reply]
And note that `trurl` can also get URLs from stdin, so if you have a file that contains those URLs, and you want to extract `{query:img_url}` for each one, you can just use:
```sh
trurl -f - -g '{query:img_url}' < baseurl > imgurls
```
Or if you want to edit the file inplace, you can use `:%!trurl -f - -g {query:img_url}` in vim. B)

**Reply by Indyandie** [reply]
Would it be a bad idea to have a way to encode a URL or parts of a URL, like query params?

**Reply by emanuele6** [reply]
@Indyandie I don't really understand what you mean, it sounds like what `trurl` already does normally. Can you open a new discussion and clarify?
